### PR TITLE
added induction heads demo, readme, and added matplotlib to requireme…

### PR DESCRIPTION
**Pull request by adil-a**
Added the Induction Heads demo, a README (unsure if I need to introduce what induction heads are), and `matplotlib` as a requirement.

### CODE REVIEW

File: demos/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# FlexModel Demos
+**Note**: If your system does not have InfiniBand, please be sure to run `export NCCL_IB_DISABLE=1` prior to running the experiments.
+## Induction Heads
+
+The file `induction_heads_multigpu.py` takes in a Llama-2 HuggingFace model path and outputs two plots which visualize the per token loss and a heatmap of the induction scores across all attention heads.
+
+Before running the file, you might need to insert `DummyModule` from `flex_model.core` in order to store the activation maps. `DummyModule` is just an `nn.Module` that is an identity function and its primary purpose is to allow us to attach a `HookFunction` at any arbitrary point during a forward pass. An example of this is below
```
Comment (MChoi-git) [inline demos/README.md:7]: minor gripe: "An example of this is [shown] below[:]". Don't need to have "
### Llama-2 Attention Module" sub-header.

File: demos/induction_heads_multigpu.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+def main(args: Namespace) -> None:
+    """Execute main demo.
+
+    Args:
+    ----
+        args: Command-line arguments
+    """
+    rank = int(os.environ.get("LOCAL_RANK", 0))
+    torch.cuda.set_device(rank)
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:227]: This can be run in the `setup()` function

File: demos/induction_heads_multigpu.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+    seq_len = 50
+    prompt = torch.randint(
+        500, 15000, (4, seq_len)).to(torch.cuda.current_device(),
+    )
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:232]: Maybe alias these numbers to make it apparent what they represent. May also be interesting to make them configurable from the command line

File: demos/induction_heads_multigpu.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+
+def args() -> Namespace:
+    """Parse command-line arguments."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_path", type=str, required=True)
+    return parser.parse_args()
+
+
+def setup_model(model_path: str, rank: int) -> tuple[nn.Module, LlamaConfig]:
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:51]: Minor: Signature would be cleaner without the rank arg, which can just be obtained via `dist.get_rank()` or other

File: demos/induction_heads_multigpu.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+        "mixed_precision": mp_policy,
+    }
+    return config
+
+
+def calculate_induction_score(
+        config: LlamaConfig,
+        activation_dict: dict[str, torch.Tensor],
+        module_names: list[str],
+        sequence_length: int,
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:125]: Would probably be better to just pass the `num_hidden_layers` and `num_attention_heads` rather than the entire `LlamaConfig`, since you're already passing the `sequence_length`

File: demos/induction_heads_multigpu.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+    Returns:
+    -------
+        A list of model names that we're applying HookFunctions to
+    """
+    name_placeholder = """_fsdp_wrapped_module.model.layers.
+        {}._fsdp_wrapped_module.self_attn.dummy"""
+    module_names = [
+        name_placeholder.format(i) for i in range(config.num_hidden_layers)
+    ]
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:183]: Use f-string instead here, ie. `module_names = [f"layer_prefix.{i}.layer_postfix" for i in range(...)]`

File: demos/induction_heads_multigpu.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+    plt.imshow(induction_score_store.cpu().numpy(), origin="lower")
+    plt.xlabel("Head")
+    plt.ylabel("Layer")
+    plt.title("Induction Score by Head")
+    plt.colorbar()
+    plt.savefig("induction_score_by_head.png", bbox_inches="tight")
+
+
+def get_module_names(config: LlamaConfig) -> list[str]:
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:168]: Config not really needed here, makes more sense to just pass the `num_hidden_layers`

File: demos/induction_heads_multigpu.py (hunk 5 above)
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:125]: I think this indentation of the function signature is one tab too long

File: demos/induction_heads_multigpu.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        attn_maps = activation_dict[module_name][0].detach().to(
+            torch.cuda.current_device(),
+        )
+        induction_stripe = attn_maps.diagonal(
+            dim1=-2, dim2=-1, offset=1-sequence_length,
+        )
+        induction_score = einops.reduce(
+            induction_stripe, "batch head_index position -> head_index", "mean",
+        )
+        induction_score_store[i, :] = induction_score
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:158]: Add comments on these describing the operations that are being performed to calculate the induction scores (or put it in the docstring for this function)

File: demos/induction_heads_multigpu.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+
+    plt.imshow(induction_score_store.cpu().numpy(), origin="lower")
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:160]: Minor: Can swap `x.cpu().numpy() -> x.numpy(force=True)`. This also consequently calls `detach()`, which you wouldn't need to call above either. Also might not be necessary to run this part on GPU

File: demos/induction_heads_multigpu.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    module_names = [
+        name_placeholder.format(i) for i in range(config.num_hidden_layers)
+    ]
+    return module_names
+
+
+def calculate_per_token_loss(
+        logits: torch.Tensor,
+        prompt: torch.Tensor,
+    ) -> None:
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:190]: indentation here is one-tab too far

File: demos/induction_heads_multigpu.py [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+    Args:
+    ----
+        logits: The model's output logits
+        prompt: The input prompt sequence
+    """
+    # Calculate per token loss
+    log_probs = F.log_softmax(logits, dim=-1)
+    predicted_log_probs = -log_probs[..., :-1, :].gather(
+        dim=-1, index=prompt[..., 1:, None],
+    )[..., 0]
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:202]: Not obvious what is going on here, maybe add comment on the operation

File: demos/induction_heads_multigpu.py [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+
+    # Average loss across the batch dimension
+    loss_by_position = einops.reduce(
+        predicted_log_probs, "batch position -> position", "mean",
+    )
+
+    plt.plot(
+        list(range(len(loss_by_position))),
+        loss_by_position.detach().cpu().numpy(),
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:211]: Again, can call `x.numpy(force=True)` instead.

File: demos/induction_heads_multigpu.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+    model = FlexModel(
+        model,
+        output_dict,
+        data_parallel_size=dist.get_world_size(),
+    )
+
+    # Register hooks for activations
+    module_names = get_module_names(config)
+    for i in range(config.num_hidden_layers):
+        module_name = module_names[i]
```
Comment (MChoi-git) [inline demos/induction_heads_multigpu.py:256]: Can just use `for mod_name in module_names: ...` instead.

**Review by MChoi-git — COMMENTED** [review]
Left some comments about some function signatures and arguments being passed. Also some comments on adding some command-line options and more descriptions of tensor operations being performed.

File: demos/induction_heads_multigpu.py (hunk 4 above)
Comment (adil-a) [reply demos/induction_heads_multigpu.py:51]: Turned this into a global

File: demos/induction_heads_multigpu.py (hunk 9 above)
Comment (adil-a) [reply demos/induction_heads_multigpu.py:160]: I think I'll keep this for now. I know it's more verbose, but I think it's ok because it's a demo and it's better to be explicit here.

**Comment by MChoi-git** [comment]
LGTM, but might need to update your `requirements.txt` to reflect the main branch
